View.elm

```diff
@@ -0,0 +1,304 @@
+
+
+onClick : msg -> View tipe custom msg -> View (OnClick msg tipe) custom msg
+onClick msg child =
+    View <| OnClick msg (toSubView child)
+
+
+text : String -> View String custom msg
+text text =
+    View <| Text text
+
+
+just : Maybe a -> View tipe1 custom msg -> (a -> View tipe2 custom msg) -> View (OneOf tipe1 tipe2) custom msg
+just maybe (View first) second =
+    case maybe of
+        Nothing ->
+            View first
+
+        Just x ->
+            let
+                (View sub) =
+                    second x
+            in
+            View sub
+
+
+list : List (View tipe custom msg) -> View (List tipe) custom msg
+list xs =
+    View <| List (List.map toSubView xs)
+
+
+tuple2 : View tipe1 custom msg -> View tipe2 custom msg -> View ( tipe1, tipe2 ) custom msg
+tuple2 child1 child2 =
+    View <| List [ toSubView child1, toSubView child2 ]
+
+
+{-| Useful during development, to get unimplemented parts of view functions compiling.
+-}
+debug : View tipe custom msg
+debug =
+    View <| Text "Debugging value. Remove me!"
+
+
+map : (msgA -> msgB) -> View tipe custom msgA -> View tipe custom msgB
+map fn (View subView) =
+    View (mapSubView fn subView)
+
+
+mapSubView : (msgA -> msgB) -> SubView custom msgA -> SubView custom msgB
+mapSubView fn subView =
+    case subView of
+        H1 child ->
+            H1 (mapSubView fn child)
+
+        H2 child ->
+            H2 (mapSubView fn child)
+
+        Section child ->
+            Section (mapSubView fn child)
+
+        P child ->
+            P (mapSubView fn child)
+
+        Text text ->
+            Text text
+
+        List children ->
+            List (List.map (mapSubView fn) children)
+
+        OnClick msg child ->
+            OnClick (fn msg) (mapSubView fn child)
+
+        Custom custom child ->
+            Custom custom (mapSubView fn child)
+
+
+
+-- # View generation
+-- One example of a view interpreter, this one producing plain Html.
+
+
+mkView : (custom -> Html msg -> Html msg) -> View tipe custom msg -> Html msg
+mkView viewCustom (View subView) =
+    mkSubView viewCustom [] subView
+
+
+mkSubView : (custom -> Html msg -> Html msg) -> List (Html.Attribute msg) -> SubView custom msg -> Html msg
+mkSubView viewCustom attrs subView =
+    case subView of
+        H1 child ->
+            Html.h1 attrs (List.map (mkSubView viewCustom []) (toChildren child))
+
+        H2 child ->
+            Html.h2 attrs (List.map (mkSubView viewCustom []) (toChildren child))
+
+        Section child ->
+            Html.section attrs (List.map (mkSubView viewCustom []) (toChildren child))
+
+        P child ->
+            Html.p attrs (List.map (mkSubView viewCustom []) (toChildren child))
+
+        Text text ->
+            Html.text text
+
+        List children ->
+            Html.div attrs (List.map (mkSubView viewCustom []) children)
+
+        OnClick msg child ->
+            mkSubView viewCustom [ Html.Events.onClick msg ] child
```

Comment:
💭 not sure if this should be a thing and not just allowing `onClick` only on `button` and `a`.

Reply:
I like the general principle behind this idea, but in practice I think there's enough examples where we're listening to clicks on other elements, that introducing this would be too big a handicap.